platform/pulpcore/app/models/content.py

```diff
@@ -1,12 +1,49 @@
 """
 Content related Django models.
 """
+from django.core import validators
 from django.core.files.storage import default_storage
 from django.db import models
 
+
 from pulpcore.app.models import Model, MasterModel, Notes, GenericKeyValueRelation
 
 
+class Artifact(Model):
+    """
+    A file associated with a piece of content.
+
+    Fields:
+
+        file (models.FileField): The stored file.
+        size (models.IntegerField): The size of the file in bytes.
+        md5 (models.CharField): The MD5 checksum of the file.
+        sha1 (models.CharField): The SHA-1 checksum of the file.
+        sha224 (models.CharField): The SHA-224 checksum of the file.
+        sha256 (models.CharField): The SHA-256 checksum of the file.
+        sha384 (models.CharField): The SHA-384 checksum of the file.
+        sha512 (models.CharField): The SHA-512 checksum of the file.
+    """
+    def storage_path(self, name):
+        """
+        Callable used by FileField to determine where the uploaded file should be stored.
+
+        Args:
+            name (str): Original name of uploaded file. It is ignored by this method because the
+                sha256 checksum is used to determine a file path instead.
+        """
+        return default_storage.get_artifact_path(self.sha256)
```

Comment:
It has been my experience that for this to work, the Artifact must be a `deconstructable`.  Otherwise, it just fails silently.  Did you validate the file is actually written to the storage location?

Also, I thought we agreed a cleaner approach would be to add a path calculator to the storage module and add as an attribute to the `FileStorage`.  See:  https://github.com/pulp/pulp/pull/3100.